### PR TITLE
Implement language switcher

### PR DESCRIPTION
**Pull request by 0140454**
Related issue: #310

### Preview
Language switcher demo

### CODE REVIEW

**Review by m2049r — CHANGES_REQUESTED** [review]
can you please put the

```
<string name="menu_language">Language</string>
<string name="language_system_default">Use System Language</string>
```

at the bottom of all strings.xml - translators then can find them easier.

File: app/src/main/java/com/m2049r/xmrwallet/LoginActivity.java [hunk 1]
```diff
@@ -959,6 +965,45 @@ void copyFile(File src, File dst) throws IOException {
...
+            @Override
+            public int compare(Locale locale1, Locale locale2) {
+                return locale1.getDisplayName(locale1).compareTo(locale2.getDisplayName(locale2));
+            }
+        });
+
+        localeDisplayName[0] = getString(R.string.language_system_default);
+        for (int i = 1; i < localeDisplayName.length; ++i) {
+            Locale locale = translatedLocales.get(i - 1);
+            localeDisplayName[i] = locale.getDisplayName(locale);
```
Comment (m2049r) [inline app/src/main/java/com/m2049r/xmrwallet/LoginActivity.java]: this results in some languages being capitalised and others not - the list doesn't look good.

File: app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Locale;
+
+public class LocaleHelper {
+    private static final String PREFERRED_LOCALE_KEY = "preferred_locale";
+    private static Locale SYSTEM_DEFAULT_LOCALE = Locale.getDefault();
+
+    public static ArrayList<Locale> getAvailableLocales(Context context) {
+        ArrayList<Locale> locales = new ArrayList<>();
+        HashSet<String> localizedStrings = new HashSet<>();
+        int compareResId = R.string.menu_about;
```
Comment (m2049r) [inline app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java]: is this really the correct string to compare? i am guessing some languages have the same string for "About" - maybe we should add a specialized string to id the translations, like ```R.string.language``` which can maybe also be used for the display in the dialog list ?

File: app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        // Application default locale
+        locales.add(Locale.ENGLISH);
+        localizedStrings.add(getLocaleString(context, Locale.ENGLISH, compareResId));
+
+        // Locale with region
+        locales.add(Locale.CHINA);
+        localizedStrings.add(getLocaleString(context, Locale.CHINA, compareResId));
+
+        locales.add(Locale.TAIWAN);
+        localizedStrings.add(getLocaleString(context, Locale.TAIWAN, compareResId));
```
Comment (m2049r) [inline app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java]: why is it necessary to add these three manually? why are they special?

File: app/src/main/java/com/m2049r/xmrwallet/LoginActivity.java [hunk 4]
```diff
@@ -959,6 +965,45 @@ void copyFile(File src, File dst) throws IOException {
...
+
+        Collections.sort(translatedLocales, new Comparator<Locale>() {
+            @Override
+            public int compare(Locale locale1, Locale locale2) {
+                return locale1.getDisplayName(locale1).compareTo(locale2.getDisplayName(locale2));
+            }
+        });
+
+        localeDisplayName[0] = getString(R.string.language_system_default);
+        for (int i = 1; i < localeDisplayName.length; ++i) {
```
Comment (m2049r) [inline app/src/main/java/com/m2049r/xmrwallet/LoginActivity.java]: persoanlly, i like ```i++``` better as it indicates we are incrementing after the loop.

**Review by m2049r — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/m2049r/xmrwallet/LoginActivity.java (hunk 1 above)
Comment (0140454) [reply app/src/main/java/com/m2049r/xmrwallet/LoginActivity.java]: I can capitalize first letter of display name, such as Language setting in System Settings.
How about this idea?

File: app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java (hunk 3 above)
Comment (0140454) [reply app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java]: When enumerating system supported locales and comparing string, system will fallback to other region which use the same language if the locale is not translated.

For instance, zh-CN, zh-HK and zh-TW are used by China, Hong Kong and Taiwan respectively.
If we don't add locale manually, "Chinese (Hong Kong)" will be listed because system returns translated string from zh-TW.

So, i add them manually to inform user that this is translated for the specific region and there may be different translated terms corresponding to the same English.
In short, just for more accurate translated terms.

File: app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java (hunk 2 above)
Comment (0140454) [reply app/src/main/java/com/m2049r/xmrwallet/util/LocaleHelper.java]: It's a good idea.

If there is no better idea and you are sure to do this, i will modify code and add resource like ```<string name="language">English</string>``` to all strings.xml

**Review by m2049r — APPROVED** [review]

**Comment by 0140454** [comment]
@m2049r 
In the current implementation, `preferred_locale` saved in the preference is incorrect.

For instance, `preferred_locale` is **af** if I change app language to English, because **af** is the first tested locale when enumerating system supported locales.
If there is **af** translation in the future, app will use it instead of English after restarting.

To fix this problem, I use Gradle script to enumerate existing **values-\*** folder as translated locales list.
And I capitalized locale display name and removed `R.string.language`.

Commit is [here](https://github.com/0140454/xmrwallet/commit/6cd138f6fa692a33bd7cbb2805a994f88c1dad24). Do you think this way is better than the current? If yes, I will open a new PR for updating.

**Comment by m2049r** [comment]
@0140454 maybe add English before everything else in ```getAvailableLocales()```?

but personally, i like your new solution better as it is simpler at runtime.

**Comment by m2049r** [comment]
> maybe add English before everything else in getAvailableLocales()?

I suggested this to avoid the **af** problem in the orginal implementation :)

**Comment by 0140454** [comment]
Oh, I misunderstood your meaning...

But although we can test English first, the same problem will happens in other locales.
For instance, ```preferred_locale``` is **zh-HK** when choosing **zh-TW**.
